worker/src/RTC/ActiveSpeakerObserver.cpp

```diff
@@ -0,0 +1,523 @@
+	constexpr uint32_t LevelsBuffLen{ LongCount * N3 * N2 };
+	constexpr double MinActivityScore{ 0.0000000001 };
+
+	inline int64_t BinomialCoefficient(int32_t n, int32_t r)
+	{
+		int32_t m = n - r;
+
+		if (r < m)
+		{
+			r = m;
+		}
+
+		int64_t t = 1;
+		for (int64_t i = n, j = 1; i > r; i--, j++)
+		{
+			t = t * i / j;
+		}
+
+		return t;
+	}
+
+	inline double ComputeActivityScore(
+	  const uint8_t vL, const uint32_t nR, const double p, const double lambda)
+	{
+		double activityScore = std::log(BinomialCoefficient(nR, vL)) + vL * std::log(p) +
+		                       (nR - vL) * std::log(1 - p) - std::log(lambda) + lambda * vL;
+
+		if (activityScore < MinActivityScore)
+		{
+			activityScore = MinActivityScore;
+		}
+
+		return activityScore;
+	}
+
+	inline bool ComputeBigs(
+	  const std::vector<uint8_t>& littles, std::vector<uint8_t>& bigs, uint8_t threashold)
+	{
+		uint32_t littleLen       = littles.size();
+		uint32_t bigLen          = bigs.size();
+		uint32_t littleLenPerBig = littleLen / bigLen;
+		bool changed             = false;
+
+		for (uint32_t b = 0, l = 0; b < bigLen; b++)
+		{
+			uint8_t sum = 0;
+
+			for (uint32_t lEnd = l + littleLenPerBig; l < lEnd; l++)
+			{
+				if (littles[l] > threashold)
+				{
+					sum++;
+				}
+			}
+			if (bigs[b] != sum)
+			{
+				bigs[b] = sum;
+				changed = true;
+			}
+		}
+		return changed;
+	}
+
+	ActiveSpeakerObserver::ActiveSpeakerObserver(const std::string& id, json& data)
+	  : RTC::RtpObserver(id)
+	{
+		MS_TRACE();
+
+		auto jsonIntervalIt = data.find("interval");
+
+		if (jsonIntervalIt == data.end() || !jsonIntervalIt->is_number())
+			MS_THROW_TYPE_ERROR("missing interval");
+
+		this->interval = jsonIntervalIt->get<int16_t>();
+
+		if (this->interval < 100)
+			this->interval = 100;
+		else if (this->interval > 5000)
+			this->interval = 5000;
+
+		this->periodicTimer = new Timer(this);
+
+		this->periodicTimer->Start(interval, interval);
+	}
+
+	ActiveSpeakerObserver::~ActiveSpeakerObserver()
+	{
+		MS_TRACE();
+
+		delete this->periodicTimer;
+	}
+
+	void ActiveSpeakerObserver::AddProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].producer = producer;
+		this->mapProducerSpeaker[producer->id].speaker  = new Speaker();
+	}
+
+	void ActiveSpeakerObserver::RemoveProducer(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (mapProducerSpeaker[producer->id].speaker != nullptr)
+		{
+			delete mapProducerSpeaker[producer->id].speaker;
+		}
+
+		this->mapProducerSpeaker.erase(producer->id);
+
+		if (producer->id == this->dominantId)
+		{
+			Update();
+		}
+	}
+
+	void ActiveSpeakerObserver::ProducerResumed(RTC::Producer* producer)
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].speaker->paused = false;
```

Comment:
This method is called by Router for every Producer that is resumed. We cannot throw here. Let's just resume the Speaker if the corresponding Producer is in the map.

```suggestion
		auto it = this->mapProducerSpeaker.find(producer->id);
		
		if (it != this->mapProducerSpeaker.end())
		{
			auto& rtpObserver = it->second;
			
			rtpObserver.speaker->paused = false;
		}
```